[ios] Cache Management API

jmkiley opened the pull request:
iOS version of https://github.com/mapbox/mapbox-gl-native/pull/14962

Tracking my progress

Core Method | iOS Method/property | implementation | iosapp example | tests | docs | benchmarking
-- | -- | -- | -- | -- | -- | --
invalidateAmbientCache() | - invalidateAmbientCache | X | X | X | X |
invalidateRegion(int64_t regionID) | - invalidateOfflineRegionForPack: | X  | X | X | X |
maximumAmbientCacheSize(uint64_t) | setMaximumAmbientCacheSize  | X | X | X | X |
clearAmbientCache() | - clearAmbientCache | X |X | X | X |  |
resetDatabase() | - resetDatabase | X | X | X | X |

To Do:
- [x] Implementation
- [x] Document the new API.
- [x] Revise existing documentation to clarify that `MGLOfflineStorage` manages both offline packs - and the ambient cache
- [x] Add a demo to `iosapp`
- [ ] Add tests
~- [ ] Benchmark/time the methods.~ Tailwork.

cc @langsmith @chloekraw

jmkiley commented:
I am currently getting this warning when I call `resetDatabase` in both `iosapp` and tests: 
https://github.com/mapbox/mapbox-gl-native/blob/24df7fe5f518dfe3dfbdb138bb12a785251d2395/platform/default/src/mbgl/storage/offline_database.cpp#L128

Not sure if I'm missing something with my implementation.

cc @tmpsantos

1ec5 commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -303,6 +303,7 @@ MGL_EXPORT
  */
 - (void)removePack:(MGLOfflinePack *)pack withCompletionHandler:(nullable MGLOfflinePackRemovalCompletionHandler)completion;
 
+- (void)invalidateOfflineRegionForPack:(MGLOfflinePack *)pack withCompletionHandler:(void (^)(NSError * _Nullable))completion;
```
An “offline region” in mbgl jargon is equivalent to an “offline pack” in iOS/macOS-speak. In the public API, “offline region” usually refers to what’s known internally as an “offline region definition”. So `-invalidatePack:completionHandler:` would be less confusing.

jmkiley replied:
Thank you!

1ec5 commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +342,29 @@ MGL_EXPORT
+ This method should be called before using the database.
+ */
+
+- (void)setMaximumAmbientCacheSize:(NSInteger)cacheSize withCallback:(void (^)(NSError *_Nullable error))completion;
```
Is there a way to get the maximum ambient cache size? It’s weird to have a setter without a getter. Also, it’s unusual for a setter to be asynchronous – does it need to be asynchronous?

jmkiley replied:
`setMaximumAllowedMapboxTiles` does not have a getter, so I'm comfortable with the lack of getter for the maximum ambient cache size. The maximum ambient cache size shouldn't be modified after the database has been created. Not sure of any use cases where a getter would be necessary. I updated the table in the initial PR to reflect this.

I was unsure about making it synchronous, since this should be called before the database is initialized. Could you explain a little more about why it should be synchronous?

julianrex replied:
> The maximum ambient cache size shouldn't be modified after the database has been created.

Is this true? What happens if it is called after - for example halving the size? Is this valid @tmpsantos?

If it is true, then this shouldn't be a setter, and we should find a more appropriate place to set this value.

julianrex replied:
To answer my own question:
> This method should always be called before using the database,
> otherwise the default maximum size will be used.

But, also

> Setting the size to 0 will disable the cache if there is no
> offline region on the database.

Does that mean that disabling the cache depending on whatever conditions is not possible?

jmkiley replied:
There is some additional discussion in https://github.com/mapbox/mapbox-gl-native/issues/15025 that may be helpful @julianrex. One of the things I'm working on is making sure the documentation around this is clear.

> Does that mean that disabling the cache depending on whatever conditions is not possible?

If this is set after the offline database has been initialized, the default cache size will be used. So conditionally setting the size after the database has been created wouldn't really work.

Another option could be setting the maximum ambient cache size in the `Info.plist`.

jmkiley replied:
Also, just a reminder that this PR is still a _draft_. Feedback is appreciated, but it is not ready for review yet 🐌

jmkiley commented:
I updated the iOS docs around `setMaximumAmbientCacheSize` to match the Android docs, which state that the maximum size should be set before the map and style have been loaded. 

I am also still seeing the issue I commented about [here](https://github.com/mapbox/mapbox-gl-native/pull/14978#issuecomment-506902462), and am not sure how to handle a warning. It doesn't seem to be caught by the exception.

jmkiley commented:
Thank you @fabian-guerra for helping me troubleshoot the issue with `-resetDatabaseWithCompletionHandler`. The issue was that I had `dispatch_async(dispatch_get_main_queue(), [&, completion, error](void) {
` instead of `dispatch_async(dispatch_get_main_queue(), ^{`. It does not seem to be crashing now 🎉 .

jmkiley commented:
I updated some of the docs and tests, but I believe that this is ready for review.

Most of the methods that I added to `iosapp` include basic timing for the methods, as well as measure the size of the cache before and after the method has been called. cc @juli

friedbunny commented on platform/ios/app/MBXAppDelegate.m:
```diff
@@ -13,7 +13,11 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     [MGLLoggingConfiguration sharedConfiguration].loggingLevel = MGLLoggingLevelFault;
 #endif
-
+    [[MGLOfflineStorage sharedOfflineStorage] setMaximumAmbientCacheSize:30000000 withCallback:^(NSError * _Nullable error) {
```
Tip: when specifying big numbers that could be difficult to read or easily mistake for a different number, make them a named variable.

(Some languages allow you to insert underscores [e.g., in place of commas] to create visual separation... unfortunately, objc is not one of those languages.)

friedbunny commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+ has been set. This handler is executed synchronously on the main queue.
+ */
+
+- (void)setMaximumAmbientCacheSize:(NSInteger)cacheSize withCallback:(void (^)(NSError *_Nullable error))completion;
```
This method currently defines `cacheSize` as `NSInteger` — is it valid to specify a negative number? If not, let’s make it `NSUInteger`.

jmkiley commented on platform/ios/app/MBXViewController.m:
```diff
@@ -1765,6 +1785,62 @@ - (NSString *)telemetryDebugLogFilePath
     return filePath;
 }
 
+- (void)invalidateAmbientCache {
```
Sounds good! I can move these to `ios-sdk-examples`.

jmkiley commented on platform/darwin/src/MGLOfflineStorage.mm:
```diff
@@ -445,6 +446,29 @@ - (void)_removePack:(MGLOfflinePack *)pack withCompletionHandler:(MGLOfflinePack
+
+- (void)invalidatePack:(MGLOfflinePack *)pack withCompletionHandler:(void (^)(NSError * _Nullable))completion {
+    
+    if (!completion) { return; }
```
Good point, thank you.

jmkiley commented:
@friedbunny I think this is ready for re-review, when you have a chance.

friedbunny commented on platform/darwin/test/MGLOfflineStorageTests.mm:
```diff
@@ -251,7 +251,7 @@ - (void)testClearCache {
         [expectation fulfill];
     }];
-    [self waitForExpectationsWithTimeout:1 handler:nil];
+    [self waitForExpectationsWithTimeout:1- handler:nil];
```
👀

jmkiley replied:
oops

jmkiley commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+ cleared. This handler is executed asynchronously on the main queue.
+ */
+
+- (void)clearAmbientCacheWithCompletionHandler:(void (^)(NSError *_Nullable error))completion;
```
I believe it is because the database is deleted, then reinitialized, while resources related to the ambient cache are deleted/cleared from the database without any reinitialization.

friedbunny commented on platform/darwin/src/MGLOfflineStorage.mm:
```diff
@@ -445,6 +443,29 @@ - (void)_removePack:(MGLOfflinePack *)pack withCompletionHandler:(MGLOfflinePack
+
+    _mbglFileSource->invalidateOfflineRegion(region, [&](std::exception_ptr exception) {
+        if (exception) {
+            error = [NSError errorWithDomain:MGLErrorDomain code:-1 userInfo:@{
```
See [`MGLErrorCode`](https://github.com/mapbox/mapbox-gl-native/blob/6d4ef2177c4a5fb4dcb869bc560145285a28c434/platform/darwin/src/MGLTypes.h#L37-L57) for existing codes (and where you’d define new ones).

chloekraw commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+ the latest version on the server are updated.
+ 
+ This is more efficient than cleaning the cache because valid local tiles will
+ not be downloaded again.
```
I'd say this is not quite accurate: this method clears the entire ambient cache by marking tiles as invalid. Currently there's no way to specify which specific tiles are valid or not; all tiles in the ambient cache are considered invalid and re-downloaded if this method is triggered.

This is more efficient than **deleting** the ambient cache or the offline database. I also recommend referencing the specific APIs for those other two methods in parentheses so it's clear that this is the most preferred method for clearing the ambient cache. 

It might even be worth making that the last line in this section of documentation:

"This is the recommended method for clearing the ambient cache. It's the most efficient and least computationally expensive method."

jmkiley replied:
Okay, sounds good. I was going off the core documentation. It sounds like the behavior doesn't match this? https://github.com/mapbox/mapbox-gl-native/blob/68e3cbacbc35ae6cd6c1e0fe2ab59ed262c32786/include/mbgl/storage/default_file_source.hpp#L196-L200

chloekraw replied:
Ahh, thanks. I'll try to sync with @tmpsantos tomorrow to clear things up. cc @zugaldia

chloekraw commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+- (void)invalidateAmbientCacheWithCompletionHandler:(void (^)(NSError *_Nullable error))completion;
+
+/**
+ Clears the ambient cache, freeing storage space. This method does not
```
Change to "Deletes the ambient cache."

Let's leave out "freeing storage space" -- developers will be more incentivized to use this method than they should be.

chloekraw replied:
Should we also mention the reinitialization of the ambient cache here?

jmkiley replied:
Per an earlier review from @friedbunny, I would like to keep it as `clears` so that we are starting the documentation with the verb from method name. How about `Clears the ambient cache by deleting resources.`?

chloekraw replied:
Ah okay, fair enough. The core documentation consistently uses "erase" as the verb but the method name does start with `clear`, and I understand the desire to keep those consistent. I'm also ok with using "erase" as a verb instead of "delete" to match the core documentation:

> Clears the ambient cache by erasing resources from the ambient cache.

FWIW I think the rest of the core documentation is good to copy and paste over to the platform side:

> This operation can be potentially slow because it will trigger a VACUUM on SQLite, forcing the database to move pages on the filesystem.

> Resources overlapping with offline regions will not be affected by this call.

julianrex replied:
Core and Android use `clearAmbientCache` (unless I'm looking at the wrong thing).

chloekraw commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+ affect resources shared with offline regions.
+ 
+ @param completion The completion handler to call once the ambient cache has been
+ cleared. This handler is executed asynchronously on the main queue.
```
Use "deleted" instead of cleared

chloekraw replied:
Seeing the comment thread with Jason below, this documentation could also read "The completion handler to call once the ambient cache has been deleted and reinitialized."

chloekraw commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+- (void)setMaximumAmbientCacheSize:(NSUInteger)cacheSize withCompletionHandler:(void (^)(NSError *_Nullable error))completion;
+
+/**
+ Invalidates the ambient cache. This method checks that the tiles in the
```
Suggestion: "Clears the ambient cache. This method clears the ambient cache by marking tiles as invalid. It checks that the tiles in the ambient cache match those from the server. If the local tiles do not match, the tiles in the ambient cache are re-downloaded."

julianrex commented on platform/darwin/src/MGLOfflineStorage.mm:
```diff
@@ -445,6 +443,29 @@ - (void)_removePack:(MGLOfflinePack *)pack withCompletionHandler:(MGLOfflinePack
+    }
+
+    _mbglFileSource->invalidateOfflineRegion(region, [&](std::exception_ptr exception) {
+        if (exception) {
```
This can move inside the `if (completion)` below.

jmkiley replied:
If I move it inside `if (completion)`, will it only happen if there is a completion handler?

julianrex replied:
Yep, but you're only creating the `error` object, and that only gets used in the completion block. (Same thing in the methods below too.)

jmkiley commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+ in order to prevent the ambient cache from being larger than the
+ specified amount.
+ 
+ @param cacheSize The maximum size in bytes for the ambient cache.
```
This is mostly for readability. `50 MB` is easier to read than `5242880 bytes`. This matches the Android docs.

jmkiley commented on platform/darwin/src/MGLOfflineStorage.h:
```diff
@@ -341,6 +355,71 @@ MGL_EXPORT
+ 
+ This method should be called before the map and map style have been loaded.
+ 
+ This method is potentially expensive, as the database will trim cached data
```
Since there are several comments/questions about things that match the core and Android docs, should this PR be blocked from going into the beta? Can I address any docs issues specific to this PR now so I can merge today, then we can discuss the issues that are cross-platform later?

zugaldia replied:
@jmkiley considering that the core and Android implementation have already landed, and you're being consistent with them, I agree we should not block this PR on those grounds and address those concerns across core, Android, and iOS as a separate PR (before the release).

chloekraw replied:
Yeah I think this is a good idea.

friedbunny commented on platform/darwin/src/MGLOfflineStorage.mm:
```diff
@@ -341,7 +340,7 @@ - (void)_addContentsOfFile:(NSString *)filePath withCompletionHandler:(void (^)(
         if (!result) {
             NSString *description = [NSString stringWithFormat:NSLocalizedStringWithDefaultValue(@"ADD_FILE_CONTENTS_FAILED_DESC", @"Foundation", nil, @"Unable to add offline packs from the file at %@.", @"User-friendly error description"), filePath];
-            error = [NSError errorWithDomain:MGLErrorDomain code:-1 userInfo:@{
+            error = [NSError errorWithDomain:MGLErrorDomain code:9 userInfo:@{
```
Thanks for adding specific error codes for these. Instead of using the numeric code directly, use the `MGLErrorCodeModifyingOfflineStorageFailed` constant you added in `MGLTypes.h`.

friedbunny commented on platform/darwin/src/MGLOfflineStorage.mm:
```diff
@@ -486,6 +507,72 @@ - (void)setMaximumAllowedMapboxTiles:(uint64_t)maximumCount {
+    _mbglFileSource->resetDatabase([&, completion](std::exception_ptr exception) {
+        NSError *error = nil;
+        if (exception) {
+            error = [NSError errorWithDomain:MGLErrorDomain code:-1 userInfo:@{
```
Update this error code to be more specific (or, failing that, the `MGLErrorCodeUnknown` constant).

friedbunny replied:
@jmkiley you seem to have got all of the other ones, except the one I commented on. 😅

jmkiley commented:
Thanks @friedbunny, @chloekraw, @julianrex, and @zugaldia!

@friedbunny That would be good. Could I leave testing bad input as tailwork to do this week?

I addressed the feedback related to this specific PR. If there is anything else, please let me know. Maybe @chloekraw can ticket out questions about docs that are consistent cross-platform, so those can be addressed later this week?
